### DISCUSSION TITLE
New Mind Map Node Layouts: Challenges and Suggestions

### DISCUSSION THREAD

**Discussion by dpolivaev** [freeplane/freeplane#892] [Announcements]
Hi everyone,

I wanted to bring up the topic of the new mind map node layouts that have been implemented recently. While the technical implementation may be sound, I've heard some feedback that it might not be as intuitive as it could be.

Currently, each layout is described by three elements: LayoutOrientation (with values AUTO, TOP_TO_BOTTOM, LEFT_TO_RIGHT), ChildrenSides (with values AUTO, TOP_OR_LEFT, BOTTOM_OR_RIGHT, BOTH_SIDES), and ChildNodesAlignment (with values AUTO, AFTER_PARENT, BY_CENTER, BEFORE_PARENT). Any options that are set to AUTO are automatically chosen based on the node side and the parent layout orientation. These automatic layouts are useful if a node is moved or if the parent layout configuration changes, but they may not be as clear to users. When a layout has AUTO options, the icon for that layout contains a gear symbol to indicate that some layout options are being calculated automatically.

Additionally, the current layout descriptions are not necessarily intuitive and are composed of the descriptions of the individual layout elements. It might be helpful to consider more descriptive names for the layouts to make them easier to understand.

Right now, these layouts are only applied to a single parent node, which makes sense in some cases. However, it would be possible to implement layouts across multiple levels, specifying which layouts should be used for the child nodes of the next one, two, or three levels. While this would increase the number of available layouts and could potentially be confusing, it might also be useful in certain situations.

Finally, there have been some comments that the current icons for the layouts are not clear enough. It might be worth revisiting the icons and considering new ones that are more intuitive.

Given all of these considerations, I think it's important that we have a discussion about which single-level and multi-level layouts should be made available and how we can make them more intuitive for users. Let's also consider the names and icons for the layouts to ensure that they are clear and easy to understand.

What do you all think? Do you have any thoughts on which layouts would be most useful or any suggestions for more descriptive names or clearer icons?

Dimitry

(Disclaimer: I wrote this post with the assistance of ChatGPT, an AI language model developed by OpenAI.)

**Comment by dpolivaev** [comment]
Current Freeplane layout icons: 
<img width="131" alt="freeplane-layout-buttons" src="https://user-images.githubusercontent.com/1558052/208247063-3f793417-5993-40ab-9fc3-2f7e35856f03.png">

Mindmapper icons:
![mindmapper-buttons](https://user-images.githubusercontent.com/1558052/208247097-a570fe2f-d165-4698-ab12-8213cdb51699.png)

**Comment by dpolivaev** [comment]
I am surprised that nobody has written anything. It looks like nobody is interested in the feature. Really?
@cesar404 @euu2021 @guoranjun2 @reregaga @abc16361 @macmarrum @Bommi46 @cqg2

**Reply by euu2021** [reply]
I saw that other users were giving a good amount of feedback in the other posts, so I was afraid of overloading with more feedback. This week I will try to make some tests and give some suggestions.

**Reply by dpolivaev** [reply]
It seems that no one has responded to my questions.

The suggestion to support the Ishikawa diagram is outside the existing capabilities because it would require additional functionality for edges and layout and is not designed for an arbitrary number of levels. While it may be implemented in the future, at the moment I need help understanding what to do with the questions I originally asked.

**Reply by euu2021** [reply]
To clarify my comment: in the past weeks, there was a good amount of feedback in #825. I imagined that those users would keep engaged in the testing, so I avoided replying here in this thread. As nothing happened, I will start testing, now.

**Reply by dpolivaev** [reply]
Current preview 1.11.1-05 considers many of the suggestions, and the questions I asked also have them in mind. Therefore I would like to continue the discussion here.

**Reply by macmarrum** [reply]
I need time to analyze the topic of new layouts. I started doing that and I found myself puzzled.

I want to see how each of the new layouts looks like in Freeplane. So far I found the time to compose the first 3:

![layout_1,2,3](https://user-images.githubusercontent.com/7197131/208529761-b1cf855f-d534-4617-8123-f7201c5e82a1.png)

Based on this example alone, I am missing the option to change the side where the edge grows from. Here it grows from the side near the root. I would like an option to grow from the far side.

**Reply by dpolivaev** [reply]
I do not understand how you mean it. Could you create a sketch?

**Reply by cesar404** [reply]
> It looks like nobody is interested in the feature. Really?

Dimitry, sorry for the delay in answering your questions. I've been extremely busy with my job lately. Off course I'm very interested in the feature. In fact, for me personally, this is the best feature implemented in recent years. I will certainly be using it extensively going forward. I've used a lot of mind mapping software and I can assure you that, now, Freeplane is by far the best of them all.Your most recent changes have resolved the issues most relevant to me. I'm sure there can still be some improvements in certain specific areas. Please give me just a little more time to do more testing and more accurately answer the specific questions you raised in this post of yours. For now, I just hope you don't remove any of these new features. If there are problems, I'm sure they can be fixed. Thanks a lot.

**Comment by sgjohnson1981** [comment]
I'm very much interested in these new layout features. Just wasn't sure how to answer the questions. Personally, I feel going the way of intuiting based on clear icons is better than worrying about intuitive written descriptions, but maybe I'm just a visual learner. Perhaps organizing them in just a few subcategories like MindMapper there is the best way to go for now?

"Do you have any thoughts on which layouts would be most useful..."

Are you saying only a limited number of the possible combinations of the three elements can be implemented? Can a user just have the option of selecting whichever of the three elements they desire?

**Reply by sgjohnson1981** [reply]
It's also the holidays, so you may not get much participation at the moment for that reason.

**Reply by dpolivaev** [reply]
I have intentionally chosen to combine the elements in the layouts to make visual representations possible.
And if want to support complex layouts defined for multiple levels  and assignable with a single click they could be defined by combining the existing layouts and added to the list.
I agree that clear icons are important. But I am just a coder. Even creating the current set of icons I went far beyond my skill set.

**Comment by sifran-github** [comment]
I played a bit with these new mind map layouts. In the beginning I was a bit skeptical, but I do see the potential it has for creating more customized and nice looking mind maps.

I didn't reply in this thread because:
1. I find it hard to completely understand the new design. (all the implementation details about LayoutOrientation, ChildrenSides, ChildNodesAlignment and how they are mapped exactly to all the icons).
2. I'm not sure what are you asking exactly? are you asking a question from a user-experience level, or from a programming level? I mean, Are you asking just "general impressions about the new feature"? or "technical comments about the inner structure including the technical terms point 1" or do you have a specific question? it is not very clear.

Anyway because of 1 (I don't completely understand the design), and because I didn't follow the last big thread about the new feature, I didn't answer here.

But again, in general I do see how this new feature can be very helpful.

**Reply by dpolivaev** [reply]
Thanks for sharing your thoughts on the new mind map feature. I appreciate your feedback and I'm glad to hear that you see the potential it has for creating more customized and nice looking mind maps.

I understand that it may be difficult to fully grasp the design and implementation details, and that you may not have followed the previous discussion on this feature. That's okay, and I want to assure you that I value your perspective as a user.

As the developer of this feature, I'm interested in gathering user feedback to help improve and refine the feature. It's important to me that any future changes we make don't break existing functionality, so early user feedback is crucial.

I understand that it may be challenging for you to provide detailed feedback if you're not completely familiar with the design and implementation. However, if you have any ideas or examples of how you envision using this new feature, I'd love to hear them. Your examples and ideas can help me better understand your needs and how the feature could be used in the future.

**Reply by sifran-github** [reply]
you are welcome chatgpt ;)

**Reply by dpolivaev** [reply]
Do you like how I use ChatGPT or rather want to get my posts without a postprocessor?

**Reply by dpolivaev** [reply]
For clarity this is the original post:  

"I ask for user feedback because after the version is released any breaking changes should be avoided.
I know that in order to answer you need to understand my way of thinking and it might be challenging. 
On the other side, you also have the possibility to share your ideas on how exactly you want to use new features so that I can learn from your examples and find the answers myself."

**Reply by sifran-github** [reply]
> Do you like how I use ChatGPT or rather want to get my posts without a postprocessor?

It is a good question. I'm not sure.

I think that sometimes it can help, but it can also make the text sounds like a slick salesman: too good, too nice and with less of a "personal touch".

Maybe other grammar tools can be better for rephrasing texts? I also need a good solution since English is not my native language.

I noticed that with programming, I need to go over and change what chatgpt says, because it can be right but it can also be wrong... probably it applies to text as well.

**Reply by dpolivaev** [reply]
Well, there is a difference: I am much better at writing and evaluating code than text.

**Reply by dpolivaev** [reply]
But it looks like I am going not to use ChatGPT as a rule too keep the posts authentic.

**Reply by sifran-github** [reply]
Yeah I understand you completely. I'm on the same boat.

For me, I might use chatgpt for some text generation activities, maybe some formal emails, but anyway at the current level of chatgpt I will probably modify his output a bit.

I probably should just use some grammar tool, just didn't find a good one.

**Reply by dpolivaev** [reply]
We could try to ask ChatGPT to fix only grammar and spelling instead of rewriting everything in the "unpersonal perfect".

**Reply by dpolivaev** [reply]
Me: Fix grammar and spelling not changing the style:

I ask for user feedback because after the version is released any breaking changes should be avoided.
I know that in order to answer you need to understand my way of thinking and it might be challenging.
On the other side, you also have the possibility to share your ideas on how exactly you want to use new features so that I can learn from your examples and find the answers myself.

AI:

I ask for user feedback because after the version is released, it is important to avoid making any breaking changes. I understand that it may be challenging for you to understand my way of thinking, but I also have the opportunity to learn from your examples and ideas about how you want to use the new features, so that I can find the answers myself.

**Reply by dpolivaev** [reply]
Not bad

**Reply by sifran-github** [reply]
I started using https://quillbot.com/grammar-check sometimes.
Not sure if its the best, but it doesn't restructure the sentences, just a simple grammar check.

**Comment by sifran-github** [comment]
> "it would be possible to implement layouts across multiple levels, specifying which layouts should be used for the child nodes of the next one, two, or three levels. While this would increase the number of available layouts and could potentially be confusing, it might also be useful in certain situations."

Just some random thoughts regarding multiple levels (if I understood correctly):
1. I understand that it can help with implementing simple features of other programs in an easy way.
2. Probably there are many combinations that we didn't even think of. (because each level is another multiplication in terms of options that can be applied). Maybe you will implement a few "combinations" and then people will ask you for more and more... because there are many many options.
3. Maybe keeping it to single levels, can represent a "simpler" logic.
4. Theoretically, a solution that changes the child's behavior can be applied in a generic way: apply style X on child level 1, style Y on child level 2 and so on. These styles can represent a different map orientation like you mentioned here, or represent just different fonts/whatever.

edit: adding point 5:
5. i wonder how the features of jump in & filtering will be affected from this. for example, if you have a multi level rule from nodes 1->2->3 and you jump in inside level 2 or filter it without showing ancestors.

**Reply by dpolivaev** [reply]
Yes, I could add a list showing which layouts are applied for descendants at levels n+1, n+2 and so on it they have no own layout assignments. But I am afraid it could be too much.

**Reply by sifran-github** [reply]
added point 5 to my comment

**Reply by dpolivaev** [reply]
The logic could be as follows:
Imagine nodes Root -> PP -> P -> N -> C.
Only nodes Root, N, and C are visible because of the filter.

- If Layout properties of N are not defined at the node itself,  up to  K ancestor nodes can be checked, where K is a fix number equal to 1, 2 or 3.

- If Layout of node P contains definition of layout for the direct children, it is applied instead of the undefined properties.
- Otherwise, if Layout of node P contains definition of layout for the +2 level children it is applied,
- Otherwise the layout of node P itself is applied.

**Reply by dpolivaev** [reply]
@sifran-github

**Reply by sifran-github** [reply]
Yeah, I guess some algo like this it can work.
I'm glad i gave you some ideas to test (jump in, filter).

In general, my comments represent my thoughts about:
- Complexity: the complexity of allowing another feature, and how it can interact with other features. (Will it break something?)
- Generic: the generic way of implementing it. (Is there a way to implement it in a way that will allow more features?)

I'm not diving into the specific details, but it seems to me that you know the solution: You know the answers to the complexity problems, and making it too generic for additional features is probably too much (like allowing styles to define their children's styles, not just the positioning of the children. but anyway it can already be done with conditional styles).

So, overall I think that it is reasonable...

**Reply by sifran-github** [reply]
but if i had to bet.... i would guess that there are some more complexities that will produce some bugs, that we didn't think about. Some intersections between features or between node rules.

**Comment by Illegit** [comment]
I'm using some of these layouts in my maps. I find them a really important tool for analyzing info e structuring nodes.
Your explanations were needed to better understand the buttons (otherwise a little bit too hermetic).
Now icons can be more explicit as in mind mapper but I understand them being not as complete as your choice of buttons. In any case, 2 Ishikawa diagrams left and right are necessary (and probably there is the need to have a list of preset structures and the possibility to create and save some created by the user). Still, there are some buttons missing as per some in your first choice list in v01 (they were substituted with the actual list instead of being just added to: see the 2 sketches)
![image](https://user-images.githubusercontent.com/62674157/208523725-23775fde-6042-4a4b-b8a3-dc18f058fb50.png)
Thank you for this feature, I really appreciate it.

**Reply by dpolivaev** [reply]
Yes, I have replaces the sketched layouts with a more compact form, otherwise, there would be just more layouts. I can bring them back, and it would require me to add 8 combinations.

**Reply by dpolivaev** [reply]
So, I am not sure to do it or not.

**Comment by euu2021** [comment]
> Given all of these considerations, I think it's important that we have a discussion about which single-level and multi-level layouts should be made available and how we can make them more intuitive for users.

I think that something that could make the experience more intuitive when there is mixing of layouts or multi-level layouts is letting the user drop the nodes on both sides, when the layout at has the option to show nodes on both sides. 

For example, when using this layout, the user probably will want to easily move notes to the top or to the bottom of the parent. But, in order to do that, he will need to Jump into the parent:
![image](https://user-images.githubusercontent.com/77707706/208529639-55354691-9f6c-44e0-ad67-db5f74d1ca6a.png)

Having to make a Jump in, in order to move the node T to the top, breaks the flow of the experience.

**Reply by euu2021** [reply]
> breaks the flow of the experience

Actually, it goes beyond breaking the flow of the experience. It makes it almost impossible for a novice user, because he will need to, somehow, know that he needs to Jump into the parent in order to move the node to the other side, in order to achieve the result shown in the layout icon.

**Reply by dpolivaev** [reply]
Actually, if you choose "both sides" you should be able to really change the side without jumping in.

**Reply by dpolivaev** [reply]
In your example you can drag and drop node T to the top of the selected node A

**Reply by euu2021** [reply]
If I drop T on the top of A, it becomes a sibling (which is the classic behavior):
![javaw_YBqtrufVft](https://user-images.githubusercontent.com/77707706/208548957-44b0bcf0-31a9-4b12-b403-8e2c0572d8a4.gif)

 To make T a top/left child of A, I need to jump into A. The result is this:
![image](https://user-images.githubusercontent.com/77707706/208549211-fc7e0c65-3b73-4ad8-b370-10fc71d83327.png)

**Reply by dpolivaev** [reply]
From the poing of view of node "test" node "a" is a child node, and placing node T on top of it makes it its child node which is the sibling of node a.

However, from the point of view of node "a", placing node T on top of it should just change its side.

Do you see any way to decide what the user actually wants?

Probably Freeplane can treat descendant nodes of node "a" and the outside nodes differently. Would it resolve this conflict well?

@euu2021

**Reply by euu2021** [reply]
Perhaps, if the visual feedback, when hovering, was divided in two parts:
![image](https://user-images.githubusercontent.com/77707706/209443166-e08a72b7-e598-4e24-88b5-4b180954bf64.png)

I mean, currently, it's one single thing:
![image](https://user-images.githubusercontent.com/77707706/209443218-f32fc8fb-8cbf-4fc8-95a3-761305d7deae.png)

And, this would only happen if the layout has two sides.

**Reply by euu2021** [reply]
> Probably Freeplane can treat descendant nodes of node "a" and the outside nodes differently. Would it resolve this conflict well?

This wouldn't be ideal, because the user wouldn't be able to drop directly to the desired side, when dragging from the outside.

**Reply by euu2021** [reply]
Another idea is to have two radio buttons showing the property left/top, right/down. And it would show up on places like the toolbar and the menu that appears when right-clicking the node. This would allow keyboard users to change the side using a shortcut.

**Reply by dpolivaev** [reply]
I have implemented some very simple logic and you can test it.
Preview 1.11.1_06

**Reply by euu2021** [reply]
I think it's excellent. It's visually intuitive.

The only situation in which I'm not sure is this: 

If I want to make node 3 a sibling of node 2:
Scenario 1: If nodes to the right of the Root: drop 3 to the left of 2:
![javaw_KAaGyD3PIc](https://user-images.githubusercontent.com/77707706/212493496-c32a8821-d7bb-4647-8834-d1b21d2bca5e.gif)

Scenario 2: If nodes to the left of the Root: also drop 3 to the left of 2:
![javaw_M2fcc2pVkZ](https://user-images.githubusercontent.com/77707706/212493503-40d82b04-608c-4741-8a70-8aa56b2f3177.gif)

I wonder if isn't it better to mirror the Scenario 2, so the drop area to make 3 a sibling is at the right of node 2. 

I think it is more intuitive, because when the drop area is at the opposite side of the Root, it gives a sensation of "create child"; and when it's in the direction of the Root it gives a sensation of "not creating a child". So, I think Scenario 1 is already ideal; and Scenario 2 could be inverted.

**Reply by euu2021** [reply]
I see that it was implemented on Freeplane-Setup-with-Java-1.11.1-alpha-pre07, and it works fine.

Now, I noticed that the same situation happens with this layout: 
![image](https://user-images.githubusercontent.com/77707706/212567697-3dc8b9a6-6d16-48c4-a4c5-c8b9f8bcc5f3.png)

So, when the branch is at the right of the root, dropping at the top creates a sibling; but when the branch is at the left of the root, dropping at the bottom creates a sibling. So, here, also, I suggest that both sides behave equally.

**Comment by euu2021** [comment]
What I find confusing about the description of the layouts is that, even when the image clearly shows the nodes, for example, at the right, the description mentions "at the bottom or at the right":

![image](https://user-images.githubusercontent.com/77707706/208536486-d2bba3e0-76fd-4113-8034-f3997484e607.png)


![image](https://user-images.githubusercontent.com/77707706/208536349-b58cfbc8-0888-4045-895a-7b565579f33a.png)

**Reply by dpolivaev** [reply]
Good catch, I can fix it.

**Reply by dpolivaev** [reply]
fixed, preview 1.11.1_06

**Comment by abc16361** [comment]
Using different background colours of those layout icons perhaps could help grouping layouts. Currently there are 28 layouts. They look similar. If we use three different background colours, for example, it would be quicker to identify top to bottom, left to right and auto layouts. It's just a thought.

**Reply by sifran-github** [reply]
I agree with the idea of some grouping, but oppose to changing the background color because it will create a confusing interface. The current consistency of the interface is really nice.

I'm also thinking about how to minimize the impact (controls size) inside the format panel, but i'm not sure how. Probably the exact functionality needs to be decided before the decisions about the interface.

**Reply by abc16361** [reply]
Yes, I agree the functionality is more important and needs to be decided first.

**Comment by macmarrum** [comment]
Here's a rendering of the vanilla layout configurations. 
Each node's text contains a description of how I understand LayoutOrientation, ChildrenSides, ChildNodesAlignment.
Each node details contains a pictogram representing the three layout descriptors.

Significant formatting properties:
- Base distance to children: 50 pt
- Distance between children: 10 pt
- Edge style: Horizontal

For now, I'm posting only the raw data and will let it speak for itself.
I will provide my comments later.


children-layout-01,02,03
![children-layout-01,02,03](https://user-images.githubusercontent.com/7197131/208995624-7e12de6e-75b6-4ed8-b326-97cecfb3a69e.svg)


children-layout-04,05,06
![children-layout-04,05,06](https://user-images.githubusercontent.com/7197131/208995633-241034f9-7746-40cd-9f70-0042062635b5.svg)


children-layout-07,08,09
![children-layout-07,08,09](https://user-images.githubusercontent.com/7197131/208995638-423ee3c2-1578-4efd-a6ad-df4b7cae0f04.svg)


children-layout-10,11,12
![children-layout-10,11,12](https://user-images.githubusercontent.com/7197131/208995648-ef70610f-f3c6-45b9-9b4e-cfcb607ec075.svg)


children-layout-13,14,15
![children-layout-13,14,15](https://user-images.githubusercontent.com/7197131/208995656-3f35c316-64f3-4aa6-a506-99e24686fc71.svg)

children-layout-16,17,18
![children-layout-16,17,18](https://user-images.githubusercontent.com/7197131/208995672-0c55e92a-ba43-4b88-83ba-4110c20aed51.svg)


children-layout-19,20,21
![children-layout-19,20,21](https://user-images.githubusercontent.com/7197131/208995677-9b490947-83e3-461b-99d2-4f815f5436ce.svg)


children-layout-22,23,24
![children-layout-22,23,24](https://user-images.githubusercontent.com/7197131/208995688-ea30d226-b562-4a00-aeae-8f0357107d20.svg)


children-layout-25,26,27
![children-layout-25,26,27](https://user-images.githubusercontent.com/7197131/208995704-c04ff4e2-8290-4ec4-98dc-16d05318af4d.svg)

**Reply by dpolivaev** [reply]
I am looking forward to see your comments

**Comment by macmarrum** [comment]
In general, I think it's great to have more layouts available.

### 1. I am missing the following layout options from 1.10.x:
    - first child with parent
    - last child with parent
    - as parent

Regarding the new layouts, I'll start by looking at those in the orientation known from 1.10.x, called TOP TO BOTTOM in 1.11.x.

### 2. Edge start and end points change depending on (parent/child) node's width
I think the points should remain the same, regardless of node's width.
They could be configurable (I'll describe the idea in another post).

`nord.mm-children-layout-fork-smoothly-minimal-01,02,03` - this is the standard in 1.10.x (when using automatic positioning)
![nord mm-children-layout-fork-smoothly-minimal-01,02,03](https://user-images.githubusercontent.com/7197131/209946203-7a8f8c89-ab06-40f1-833a-59beb813bae1.png)

`nord.mm-children-layout-fork-smoothly-01,02,03`
![nord mm-children-layout-fork-smoothly-01,02,03](https://user-images.githubusercontent.com/7197131/209946217-6a94d849-af33-4349-9261-169b0f4f0d9f.png)


### 3. Edge start point changes depending on edge type
As in 1.10.x, when using automatic positioning, it should be the same start point irrespective of edge type

`nord.mm-children-layout-bubble-smoothly-minimal-25,26,27.png`
![nord mm-children-layout-bubble-smoothly-minimal-25,26,27](https://user-images.githubusercontent.com/7197131/209950761-0e7a2232-df4c-4853-9105-634dd2a1af0d.png)

`nord.mm-children-layout-bubble-horizontal-minimal-25,26,27.png`
![nord mm-children-layout-bubble-horizontal-minimal-25,26,27](https://user-images.githubusercontent.com/7197131/209950768-23728a1f-c778-4ddf-a2c4-49b6564c6de9.png)


### 4. Edges cross nodes or other edges 
I think the 1.10.x positioning principles should be kept, i.e. automatically-positioned nodes are never crossed by edges or other nodes.

`nord.mm-children-layout-bubble-smoothly-01,02,03.png`
![nord mm-children-layout-bubble-smoothly-01,02,03](https://user-images.githubusercontent.com/7197131/209948666-b8c05698-bb15-4acf-9a59-2cc76451e84a.png)

`nord.mm-children-layout-bubble-smoothly-minimal-01,02,03_`
![nord mm-children-layout-bubble-smoothly-minimal-01,02,03_](https://user-images.githubusercontent.com/7197131/209958158-2e637d3d-01c2-44b5-9078-20811917a559.svg)


And when a layout abandons this fundamental automatic-positioning principle, there should be a clear indication so that User is warned and can make an informed decision.

**Comment by macmarrum** [comment]
### Configurable edge start point, edge end point

![configurable_edge_start_point_and_end_point](https://user-images.githubusercontent.com/7197131/209958717-a6f147cd-d9cb-4977-aa37-fd4f96b8bbef.png)

edge start point: near child
edge end point: near parent
![p_near-c_near](https://user-images.githubusercontent.com/7197131/209959488-38256f99-14fc-435b-b1a9-35937d57d033.png)

edge start point: far from child
edge end point: near parent
![p_far-c-near](https://user-images.githubusercontent.com/7197131/209959528-b7279423-861e-4d58-9bcb-5f796989ab48.png)

**Reply by sifran-github** [reply]
it can be nice to define the location of the edge start/end.
each line has 2 attributes, LINE START and LINE END, and each of these can be just: UP/DOWN/LEFT/RIGHT.
this will allow all the combinations.

**Comment by sifran-github** [comment]
I got to test the new features using a map design that i will probably use.
I like this layout, it is nice and useful. It allows having a lot of information in one place (categories, subcategories... and many lists...)

My impressions and notes:
1. Format -> copy style & paste style - They must also include the node layout. otherwise it is a nightmare to setup all the nodes.
2. Format ->preserve format for new sibling/child nodes - also, must include node layout as well (specially for sibling...).
3. Regarding jump in/out (not really related to the new design): if i select a folded node, jump in, jump out, the node becomes unfolded. is it intended design? because I can see the benefit of having many nodes folded and jumping into them when I want to edit their content, then jumping out. but this bug is a bit annoying, and I need to manually fold the node again every time.
4. keys: (i will write another reply).
5. For some reason, if i change the "default" edge color using the style editor (ctrl+f11), it doesn't affect the nodes. I'm not sure if its the new version problem or not.

-----

![test_new_version](https://user-images.githubusercontent.com/5730355/210216348-e2bf4a6f-4f40-4451-843c-447e91d08269.png)
[test_new_version.zip](https://github.com/freeplane/freeplane/files/10330892/test_new_version.zip)

**Reply by sifran-github** [reply]
![image](https://user-images.githubusercontent.com/5730355/210229343-cacf8409-db3a-4011-abf4-f26e7cfd19c1.png)

regarding point 5, not sure why its not updating (i clicked OK...)

**Reply by dpolivaev** [reply]
1, 2: bug, fixed.
3: intentional behavior, I can change it. But before I do please start a separate discussion to see what other people think. 
5: I do not understand where exactly you make this change and which value "Automatic edge color" box has in your map.

<img width="140" alt="edgecolor" src="https://user-images.githubusercontent.com/1558052/211197892-5a444ce0-0f21-4e6d-94e1-f4d1d742d978.png">

**Reply by sifran-github** [reply]
3. I created a discussion for it.
5. Disabling "Automatic edge color" solved the issue. I always disable it, I forgot to do so when I tested the new version.

**Comment by sifran-github** [comment]
At the moment i find that the keyboard behavior is very confusing.
maybe it is not configured right...

**_SOLUTION_**
Anyway, I see 2 types of key mapping as a solution:
1. BY FUNCTION: specific keys to "go to parent/child" and "go to prev/next siblings".
2. BY DIRECTION: keys will match their directions depending on the map layout.

Personally I think that i will use option 1 and i will configure left/right to be parent/child and up/down to be siblings.
Yes, I know that in horizontal layouts, left/right will actually go "up" on the screen (because parent-children are verticals), and up/down will actually go "left" and "right" (because siblings are horizontals), but I think that i prefer it this way because it is consistent, and i don't have to think about the direction but by the function that i want to perform.

But i guess some people will find it confusing, that's why i suggest that it will be configurable.

--------
**_EXAMPLES OF THE PROBLEM_**
With that being said, here are some strange behaviors that happen currently:

1. pressing up goes here:
![image](https://user-images.githubusercontent.com/5730355/210218807-1653f4f5-1f55-44e5-9ca7-3f7970e73a66.png)
EDIT: ACTUALLY... On second thought... this is OK. it does behave functionally just like it should (go to the previous sibling, but since it is the first sibling, it goes to the end of the previous group).

2. pressing down or right goes to the child... but no key can go to the next sibling...
![image](https://user-images.githubusercontent.com/5730355/210219013-5b8ab138-1063-4909-a4b4-c8f4e35577ea.png)
Again, In this scenario I personally prefer to keep "right" key to go to the child (selection will go down) and the "down" key to go to the next sibling (the selection will go right). I know it might sound confusing, but for me it is perfect and I can remember what every key does based on the FUNCTION that it supposed to do.

3. Pressing down goes to a child...
![image](https://user-images.githubusercontent.com/5730355/210275929-fc2a35c5-d317-4785-b1b8-05e040a3ddca.png)

The down key should not behave like this. It should go to the next sibling, not into the child. It confused me and made me write point 1 (which on second thought is OK). As i mentioned in my solution, I don't like that the up/down change their behavior based on the layout, it creates a confusing interface.

The way the keys behave right now in different node layouts, If i had to guess where the selection will go when i press on up/down/left/right, I wouldn't be able to do it. There are no simple rules for it (like i suggested), and it feels like it is jumping around :/

**Comment by sifran-github** [comment]
Regarding gears layouts: I read the original post, but i still don't understand the difference between these options for example:

![image](https://user-images.githubusercontent.com/5730355/210230481-6af24fd4-3866-4a52-aac5-2cda43454121.png)

**Reply by bepolymathe** [reply]
Yes, I was asking myself the same question yesterday

**Reply by euu2021** [reply]
I did not fully grasp it, yet. But, look at this example:

Here, every node is set to the default automatic layout:
![image](https://user-images.githubusercontent.com/77707706/212496990-9432c2c3-6997-46b1-a745-ec7ffa14fb39.png)

Now, if I change node 3 to the bottom layout, this happens:
![image](https://user-images.githubusercontent.com/77707706/212497029-e5208343-692d-4dd6-bf86-8acd5ab14c4a.png)

Notice that  node 4 also adopted a bottom layout, even though I did not change it. It assumed that the user wants to have his nodes going to the bottom, even though he only changed the layout of node 3. I think it's a reasonable assumption, and a good default behavior, because, otherwise, the user would need to change the layout of every new node under this branch (or change the default for the style,  which is not easy for novice users).

On the other hand, if, from the start, every node had a simple (not smart) "left and right" layout, this would have happened:
![image](https://user-images.githubusercontent.com/77707706/212497143-45e082a7-a86c-49a3-9f4b-7e56eae51e47.png)

**Comment by euu2021** [comment]
> Finally, there have been some comments that the current icons for the layouts are not clear enough. It might be worth revisiting the icons and considering new ones that are more intuitive.

Idea to tame the complexity: make the AUTO as a separated checkbox. So, when looking at the icons, the user can focus on choosing where he wants the nodes growing to, and, after that decision, he can decide if he wants the AUTO option turned on. 
In other words, instead of having those two, he would have only the upper one:
![image](https://user-images.githubusercontent.com/77707706/212498926-4e7e29ab-6119-492c-9301-f021549c61fc.png)

At least those guys could disappear: 
![image](https://user-images.githubusercontent.com/77707706/212499346-34013cbe-3d45-499b-b41e-4810145aba77.png)

I'm not sure what to do with those, though:
![image](https://user-images.githubusercontent.com/77707706/212499383-d69ed513-2a26-4173-b16d-5c64c62948ce.png)
